### DISCUSSION TITLE
How do I run code on Server Startup?

### DISCUSSION THREAD

**Discussion by justwiebe** [nuxt/nuxt#19033] [Questions] [answered]
I'm creating a module to register the Nuxt app with a Spring Eureka discovery server, and I'm able to use several different hooks in Development mode to do this (`listen` and `ready` were working), but I can't find any that are called when the server is starting up.

```js
export default defineNuxtModule({
  meta: {
    compatibility: { nuxt: '^3.0.0' }
  },
  setup (options, nuxt) {
    let eureka;

    // Not called in Production
    nuxt.hook('listen', (server, listener) => {
      eureka = createEurekaClient(eurekaOptions);
      console.warn('Starting Eureka Heartbeat', eurekaOptions);
      eureka.start();
    });

    // Not called in Production
    nuxt.hook('close', () => {
      console.warn('Stopping eureka heartbeat');
      eureka?.stop(() => (eureka = null));
    });
  }
});

**Comment by Baroshem** [comment]
Hey, 

What about using one of the hooks listed [here](https://nuxt.com/docs/api/advanced/hooks) like `vite:serverCreated` for example?

**Reply by justwiebe** [reply]
Unfortunately it looks like that is a `build` hook

**Comment by TheAlexLichter** [comment]
I'd probably root for a nitro plugin here 🤔

**Reply by justwiebe** [reply]
Is there a list of hooks somewhere? Or a recommended path you'd suggest? Nitro docs are very lacking, lol https://nitro.unjs.io/config#hooks

**Reply by TheAlexLichter** [reply]
See [the Nuxt docs](https://nuxt.com/docs/guide/going-further/hooks#usage-with-plugins) + possible `nitro:init` hook ☺️

**Reply by justwiebe** [reply]
Thanks for the replies! I just tested, and it looks like `nitro:init` is only run on `build`. When I start up the server in production mode, it isn't run. The Nuxt docs mention `render:response` and `render:html` as being the only 2 `nitro` `runtime` hooks. Is that true?

**Reply by TheAlexLichter** [reply]
Hm, then use no hook at all and run what you want inside `defineNitroPlugin` ☺️

**Reply by justwiebe** [reply]
That's working, thanks!

**Reply by justwiebe** [reply]
@manniL actually I'm running into another problem. I was dynamically setting `runtimeConfig` values on startup by setting `process.env` variables based on results from a ConfigServer. Now that I'm doing that inside the Nitro config, it looks like I'm too late, as the runtimeConfig isn't changing after I set the environment variables. Do you know of any way to update the `runtimeConfig`?

**Comment by justwiebe** [comment, marked as answer]
Thanks to @manniL for the help! This is what I ended up with:

Module file:
```js
import { defineNuxtModule, createResolver, addServerHandler } from '@nuxt/kit';

const { resolve } = createResolver(import.meta.url);

export default defineNuxtModule({
  meta: {
    compatibility: { nuxt: '^3.0.0' }
  },
  /**
   * @param options
   * @param nuxt {Nuxt}
   */
  async setup (options, nuxt) {
    nuxt.hook('nitro:init', nitro => {
      nitro.options.plugins.push(resolve('./nitro-plugin.js'));
    });

    addServerHandler({ route: '/actuator', handler: resolve('./_info.js') });
    addServerHandler({ route: '/actuator/info', handler: resolve('./_info.js') });
    addServerHandler({ route: '/actuator/health', handler: resolve('./_health.js') });
    addServerHandler({ route: '/actuator/deregister', handler: resolve('./_deregister.js') });
  }
});
```

nitro-plugin.js
```js
import { defineNitroPlugin } from 'nitropack/runtime/plugin';
import { fetchConfigs } from './config-client.js';
import { registerEureka } from './eureka-client.js';
import { getProperty } from '../../variables.js';

export default defineNitroPlugin(async (nitroApp) => {
  await fetchConfigs();

  if (getProperty('eureka.client.register-with-eureka') === 'true') {
    registerEureka();
  } else {
    console.warn('Not registering with Eureka due to configuration');
  }
});
```

**Reply by DesertCookie** [reply]
Attempting this with Nuxt 3.12.4 yields this error: `RollupError: server/plugins/my-module.ts (1:27): "createResolver" is not exported by "node_modules/unenv/runtime/mock/proxy.mjs", imported by "server/plugins/my-module.ts". ` Which version did you test with and got this to work?

**Reply by TheAlexLichter** [reply]
@DesertCookie wrong import?

**Reply by fredericDelaporte** [reply]
For my case, just dropping a nitro plugin file in the `./server/plugins` directory was enough. I want to enable HTTP/2 for server side fetches and had a hard time finding where to do that without executing it at each request. It seems to work in development at least, I have yet to check that on an actual server.

So here is the plugin I have put in `./server/plugins`:

```
// Requires having installed the undici package.
import { setGlobalDispatcher, Agent } from 'undici'

export default defineNitroPlugin(() => {
  // Enable HTTP/2 for fetch executed on the server.
  setGlobalDispatcher(new Agent({
    allowH2: true
  }))
})
```

Now I see requests from SSR to my external APIs done with HTTP/2. But maybe I am missing something.
